### PR TITLE
GEOPY-760: Create spatial tiling estimator

### PR DESCRIPTION
**Pull request by domfournier**
**GEOPY-760 - Create spatial tiling estimator**

### CODE REVIEW

File: simpeg_drivers/utils/tile_estimate.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+                "Input must be a path to a UI JSON file or an InputFile instance."
+            )
+
+        self.params = TileParameters.build(input_file)
+
+    def start(self) -> SimPEGGroup:
+        """
+        Run the tile estimator.
+        """
+        driver = simpeg_group_to_driver(self.params.simulation, self.params.geoh5)
```
Comment (benk-mira) [inline simpeg_drivers/utils/tile_estimate.py:56]: Make driver, mesh, locations and active_cells properties to facilitate breaking start function into smaller pieces

File: simpeg_drivers/utils/tile_estimate.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        mesh = driver.inversion_mesh.mesh
+        locations = driver.inversion_data.locations
+        active_cells = active_from_xyz(
+            driver.inversion_mesh.entity, locations, method="nearest"
+        )
+        results = {}
+        counts = fibonacci_series(15)[2:]
+        for count in tqdm(counts, desc="Estimating tiles:"):
+            tiles = tile_locations(
```
Comment (benk-mira) [inline simpeg_drivers/utils/tile_estimate.py:65]: Can we break this into smaller pieces?  

Something like:

``` 
for count in tqdm():
    results[count] = self._get_result()

new_out_group = get_new_out_group(results)
return new_out_group
```
with private helper methods for lines 71 and 83 just to explain what they do in the docstring/typing and another helper for the `if params.render_plot:` block.

File: simpeg_drivers/utils/tile_estimate.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            )
+            driver.params.tile_spatial = int(count)
+            local_sim, local_map = driver.inversion_data.simulation(
+                mesh,
+                active_cells,
+                survey,
+                padding_cells=driver.params.padding_cells,
+                tile_id=count,
+            )
+            results[count] = float(survey.nD) * local_map.shape[0] * count * 8 * 1e-9
```
Comment (benk-mira) [inline simpeg_drivers/utils/tile_estimate.py:83]: what do the `* 8` and `* 1e-9` represent?  Maybe assign them to a well named variable to make it clear?

File: simpeg_drivers/utils/tile_estimate.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+from simpeg_drivers import assets_path
+from simpeg_drivers.driver import DRIVER_MAP, InversionDriver
+from simpeg_drivers.utils.utils import (
+    active_from_xyz,
+    simpeg_group_to_driver,
+    tile_locations,
+)
+
+
+class TileEstimator:
```
Comment (benk-mira) [inline simpeg_drivers/utils/tile_estimate.py:40]: I think this one could use a class level docstring with a one-liner for 'what does it do' and a larger paragraph to briefly describe 'how it does that'.

File: simpeg_drivers/utils/tile_estimate.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+            new_out_group.add_file(path)
+            figure.savefig(path)
+
+        if self.params.monitoring_directory is not None:
+            monitored_directory_copy(self.params.monitoring_directory, new_out_group)
+
+        return new_out_group
+
+    @staticmethod
+    def estimate_optimal_tile(results: dict) -> int:
```
Comment (benk-mira) [inline simpeg_drivers/utils/tile_estimate.py:111]: type the inside of the dict.  Is it `dict[int, int]` ?

File: simpeg_drivers/utils/tile_estimate.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        """
+        Estimate the optimal number of tiles to use using a circle fit.
```
Comment (benk-mira) [inline simpeg_drivers/utils/tile_estimate.py:113]: Help me out here!  Describe what is being fit to a circle and why it's optimal.

File: simpeg_drivers/utils/tile_estimate.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+        if self.params.out_group is not None:
+            new_out_group.parent = self.params.out_group
+
+        if self.params.render_plot:
+            figure = self.plot(results, locations, optimal)
+            path = self.params.geoh5.h5file.parent / "tile_estimator.png"
+            figure.savefig(path)
+            new_out_group.add_file(path)
+            figure.savefig(path)
```
Comment (benk-mira) [inline simpeg_drivers/utils/tile_estimate.py:103]: did you mean to add the second savefig?

File: simpeg_drivers/utils/tile_estimate.py (hunk 5 above)
Comment (benk-mira) [inline simpeg_drivers/utils/tile_estimate.py:111]: Can you design a unit test for this?

File: simpeg_drivers/utils/tile_estimate.py (hunk 2 above)
Comment (benk-mira) [reply simpeg_drivers/utils/tile_estimate.py:65]: ..and unit tests please!

**Review by benk-mira — CHANGES_REQUESTED** [review]
Looks cool, but I couldn't quite decipher what was going on in a few places.  I have some suggestions to update docstring and refactor for readability and unit testing.  Also let's try to re-use the BaseDriver class.

File: simpeg_drivers/utils/tile_estimate.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+        Run the tile estimator.
+        """
+        driver = simpeg_group_to_driver(self.params.simulation, self.params.geoh5)
+        mesh = driver.inversion_mesh.mesh
+        locations = driver.inversion_data.locations
+        active_cells = active_from_xyz(
+            driver.inversion_mesh.entity, locations, method="nearest"
+        )
+        results = {}
+        counts = fibonacci_series(15)[2:]
```
[the remark this answers is not in the document]
Comment (domfournier) [reply simpeg_drivers/utils/tile_estimate.py:63]: Just a way to get progressively larger tile numbers

File: simpeg_drivers/utils/tile_estimate.py (hunk 3 above)
Comment (domfournier) [reply simpeg_drivers/utils/tile_estimate.py:83]: How to convert 32 bit float array to Gb

File: simpeg_drivers/utils/tile_estimate.py (hunk 6 above)
Comment (domfournier) [reply simpeg_drivers/utils/tile_estimate.py:113]: Point of max curvature in curve marks the point where we don't gain much in reducing the problem size, but start increasing the amount of work for many more tiles. So of an estimate for the point of diminishing return.

File: simpeg_drivers/utils/tile_estimate.py (hunk 1 above)
Comment (domfournier) [reply simpeg_drivers/utils/tile_estimate.py]: Done

File: simpeg_drivers/utils/tile_estimate.py (hunk 3 above)
Comment (benk-mira) [reply simpeg_drivers/utils/tile_estimate.py]: oh, right, thanks.  Let's add that to a docstring in the refactor.

**Review by benk-mira — CHANGES_REQUESTED** [review]

**Review by benk-mira — DISMISSED** [review]
Thanks

**Comment by benk-mira** [comment]
Can we support Python Plugin Groups?
image

**Comment by benk-mira** [comment]
Any way to edit the options of the original SimPEG group instead of copying and editing the n_tiles option?
image

**Comment by benk-mira** [comment]
Let's format those floats for readability and describe the units.  Something like "Computed tile sizes (Gb) per number of tile."
image

**Comment by benk-mira** [comment]
Would be cool to format that dictionary in a table.  Check this package:
https://pypi.org/project/tabulate/

**Comment by domfournier** [comment]
> Can we support Python Plugin Groups? 

Nope, needs to a SimpegGroup.

**Pull request merged**